sklearn/utils/estimator_checks.py

```diff
@@ -625,9 +625,14 @@ def check_dtype_object(name, estimator_orig):
         if "Unknown label type" not in str(e):
             raise
 
-    X[0, 0] = {'foo': 'bar'}
-    msg = "argument must be a string or a number"
-    assert_raises_regex(TypeError, msg, estimator.fit, X, y)
+    if name not in SUPPORT_STRING:
+        X[0, 0] = {'foo': 'bar'}
+        msg = "argument must be a string or a number"
+        assert_raises_regex(TypeError, msg, estimator.fit, X, y)
+    else:
+        # If the estimator support strings passed as object dtype,
```

Comment:
Actually `X[0, 0] = {'foo': 'bar'}` is not a string entry, but a dict entry. So I think the test was correct. (Otherwise it would have been needed to modify it for the support of strings in the SimpleImputer).

Reply:
The error is also not raised for SimpleImputer with the strategy supporting mean. The difference is that the common tests are using default parameter which is the mean strategy for the SimpleImputer.

```python
In [9]: imputer = SimpleImputer(strategy='most_frequent') 
In [11]: X = np.array([[{'foo': 'bar'}, 'b', 'c']], dtype=object)
In [12]: imputer.fit(X)                                                              
Out[12]: 
SimpleImputer(copy=True, fill_value=None, missing_values=nan,
              strategy='most_frequent', verbose=0)

```

I have to check what is going wrong in check array which is the one that should raise this error.

Reply:
The error is raised by `np.asarray` only when we make the conversion. So it means that we neeed to raise such error when `dtype=object`.

Reply:
> I think `check_array` needs an update to properly deal with non-numeric input

See also https://github.com/scikit-learn/scikit-learn/issues/12148